components/codec/src/byte.rs

```diff
@@ -869,6 +943,97 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_memcmp_encode_all_in_place() {
+        // Checks whether encoded result matches expectation.
+
+        let cases: Vec<(Vec<u8>, Vec<u8>, Vec<u8>)> = vec![
+            (
+                vec![],
+                vec![0, 0, 0, 0, 0, 0, 0, 0, 247],
+                vec![255, 255, 255, 255, 255, 255, 255, 255, 8],
+            ),
+            (
+                vec![0],
+                vec![0, 0, 0, 0, 0, 0, 0, 0, 248],
+                vec![255, 255, 255, 255, 255, 255, 255, 255, 7],
+            ),
+            (
+                vec![1, 2, 3],
+                vec![1, 2, 3, 0, 0, 0, 0, 0, 250],
+                vec![254, 253, 252, 255, 255, 255, 255, 255, 5],
+            ),
+            (
+                vec![1, 2, 3, 0],
+                vec![1, 2, 3, 0, 0, 0, 0, 0, 251],
+                vec![254, 253, 252, 255, 255, 255, 255, 255, 4],
+            ),
+            (
+                vec![1, 2, 3, 4, 5, 6, 7],
+                vec![1, 2, 3, 4, 5, 6, 7, 0, 254],
+                vec![254, 253, 252, 251, 250, 249, 248, 255, 1],
+            ),
+            (
+                vec![0, 0, 0, 0, 0, 0, 0, 0],
+                vec![0, 0, 0, 0, 0, 0, 0, 0, 255, 0, 0, 0, 0, 0, 0, 0, 0, 247],
+                vec![
+                    255, 255, 255, 255, 255, 255, 255, 255, 0, 255, 255, 255, 255, 255, 255, 255,
+                    255, 8,
+                ],
+            ),
+            (
+                vec![1, 2, 3, 4, 5, 6, 7, 8],
+                vec![1, 2, 3, 4, 5, 6, 7, 8, 255, 0, 0, 0, 0, 0, 0, 0, 0, 247],
+                vec![
+                    254, 253, 252, 251, 250, 249, 248, 247, 0, 255, 255, 255, 255, 255, 255, 255,
+                    255, 8,
+                ],
+            ),
+            (
+                vec![1, 2, 3, 4, 5, 6, 7, 8, 9],
+                vec![1, 2, 3, 4, 5, 6, 7, 8, 255, 9, 0, 0, 0, 0, 0, 0, 0, 248],
+                vec![
+                    254, 253, 252, 251, 250, 249, 248, 247, 0, 246, 255, 255, 255, 255, 255, 255,
+                    255, 7,
+                ],
+            ),
+        ];
+
+        for (src, expect_encoded_asc, expect_encoded_desc) in cases {
+            let encoded_len = MemComparableByteCodec::encoded_len(src.len());
+
+            // Test encode ascending
+            let mut buffer: Vec<u8> = src.clone();
+            buffer.reserve_exact(encoded_len);
+            let output_len = MemComparableByteCodec::encode_all_in_place(&mut buffer);
+            assert_eq!(output_len, encoded_len);
+            assert_eq!(output_len, expect_encoded_asc.len());
+            assert_eq!(buffer.len(), src.len());
+            unsafe {
+                buffer.set_len(output_len);
```

Comment:
The caller must set Vector length manually, is it OK?

Reply:
I think first we should either check the length of  `src`  in `encode_all_in_place` (src_len >= dest_len),  or mark this function as `unsafe`, add `Safety` section to the docs.

Reply:
> check the length of `src` in `encode_all_in_place` (src_len >= dest_len)
I choose to use the length of `src` to indicate the length of bytes before codec. Or I can safely resize the vector to make sure the src_len >= dest_len, but must provide another parameter to indicate the length of bytes before codec, as

Reply:
I'm OK with this.  What' your opinion @breeswish ?

Reply:
I would prefer to add an extra parameter indicating how many bytes to encode in the src (and ensure that total length of src is sufficient to hold dest), to minimize the unsafety.

Reply:
Got it.